quix-frontend/client/src/state-components/users/UsersComponent.tsx

```diff
@@ -1,70 +1,155 @@
-import * as React from 'react';
+import React, {useEffect} from 'react';
+import _ from 'lodash';
 import {IUser} from '@wix/quix-shared';
-import {Table} from '../../lib/ui/components/Table';
+import {Highlighter} from '../../lib/ui/components/Highlighter';
+import {Table} from '../../lib/ui/components/table/Table';
+import {Title} from '../../lib/ui/components/table/Title';
+import {useViewState} from '../../services/hooks';
 import {usersTableFields} from './users-table-fields';
+import makePagination from '../../lib/ui/components/hoc/makePagination';
+import Input from '../../lib/ui/components/Input';
+import {FilterInitialState, InitialState, EmptyState, ErrorState} from '../../lib/ui/components/table/states';
+import {debounceAsync} from '../../utils';
 
 export interface UsersProps {
   users: IUser[];
   error: {message: string};
   onUserClicked(user: IUser): void;
 }
 
+export const CHUNK_SIZE = 100;
+
+const PaginatedTable = makePagination(Table);
+
+const search = debounceAsync((loadMore, { users, emailFilter }) => {
+  return new Promise(res => 
+    res(users?.filter(user => user.email.includes(emailFilter)) || [])
+  );
+});
+
+const States = [
+  'Initial',
+  'Error',
+  'Empty',
+  'Content',
+  'FilterInitial',
+];
+
 export function Users(props: UsersProps) {
-  const {users, error, onUserClicked} = props;
+  const {users: serverUsers, error, onUserClicked} = props;
+  const [stateData, viewState] = useViewState(States, {
+    users: [],
+    size: 0,
+    emailFilter: '',
+    errorMessage: '',
+  });
 
-  const displayLoadingState = () => (
-    <div className="bi-empty-state--loading bi-fade-in">
-      <div className="bi-empty-state-content">Loading users...</div>
-    </div>
+  useEffect(() => {
+    if (error) {
+      viewState.set('Error', { errorMessage: error.message });
+    }
+  }, [error]);
+
+  useEffect(() => {
+    if (!error) {
+      viewState.set('Initial', {users: serverUsers});
+    }
+  },[serverUsers]);
+
+  useEffect(() => {
+    if (viewState.get() !== 'Error') {
+      loadMore(0, CHUNK_SIZE + 1)(res => {
+        if (!_.isEqual(res, stateData.users) || viewState.is('Initial')) {
+          viewState.set(res.length > 0 ? 'Content' : 'Empty', {users: res});
+        } else if (stateData.users?.length > 0 && !viewState.is('Content')) {
+          viewState.set('Content');
+        }
+      });
+    }
+  }, [stateData.emailFilter, stateData.users]);
+
+  const highlightQuery = (columnName: string) => (term: string) => {
+    const text = term.replace(/\s+/g,' ');
+    
+    if (columnName === 'email') {
+      return <Highlighter
+        term={term}
+        filter={stateData.emailFilter}
+      />;
+    }
+
+    return text;
+  }
+
+  const loadMore = (offset: number, limit: number) => {
```

Comment:
these arguments are passed but are not used?

Reply:
yes, for future async pagination

Reply:
in this case, how does load more works?